### PR TITLE
PPU LLVM/Interpreter: Accurate vector instruction NaNs

### PR DESCRIPTION
**Pull request by VelocityRa**
Tested with https://github.com/RPCS3/ps3autotests/tree/master/tests/cpu/ppu_vpu.

Results in that test improved:
**In the LLVM recompiler**: by about half, ~11k different lines from realhw to ~6k.
**In the precise interpreter**: From 2746 to 353 different lines.

Needs testing to determine performance regressions and improved game compatibility.

**Note for testers:** Clear PPU game cache and firmware cache

### CODE REVIEW

File: rpcs3/Emu/Cell/PPUTranslator.cpp [hunk 1]
```diff
@@ -218,6 +222,17 @@ Function* PPUTranslator::Translate(const ppu_function& info)
 	return m_function;
 }
 
+Value* PPUTranslator::HandleNan(Value* val)
+{
+	const auto type = val->getType();
+	const auto const0_vec = ConstantVector::getSplat(type->getVectorNumElements(), ConstantFP::get(type->getVectorElementType(), 0.0));
+	const auto is_nan = m_ir->CreateFCmpUNO(val, const0_vec);
```
Comment (elad335) [inline rpcs3/Emu/Cell/PPUTranslator.cpp:229]:
```suggestion
	const auto is_nan = m_ir->CreateFCmpUNO(val, val);
```

File: rpcs3/Emu/Cell/PPUTranslator.h [hunk 2]
```diff
@@ -99,6 +101,14 @@ class PPUTranslator final : public cpu_translator
 		return SetVr(vr, expr.eval(m_ir));
 	}
 
+	llvm::Value* HandleNan(llvm::Value* val);
+
+	template <typename T>
+	auto handle_nan(T&& expr)
```
Comment (elad335) [inline rpcs3/Emu/Cell/PPUTranslator.h:107]: It's name should be signifying that it only applies to the vector instructions.

File: rpcs3/Emu/Cell/PPUInterpreter.cpp [hunk 3]
```diff
@@ -358,6 +358,29 @@ class ppu_scale_table_t
...
+
+__m128 vec_handle_nan(__m128 val)
+{
+	const auto mask_nan = _mm_cmpunord_ps(val, val);
+	const auto mask_not_nan = _mm_cmpord_ps(val, val);
+
+	const auto nans = _mm_castsi128_ps(_mm_set1_epi32(ppu_nan_u32));
+
+	const auto nan_part = _mm_and_ps(mask_nan, nans);
+	const auto not_nan_part = _mm_and_ps(mask_not_nan, val);
```
Comment (elad335) [inline rpcs3/Emu/Cell/PPUInterpreter.cpp:372]: vector select can be optimized into `result = (nans & is_nan) | (val & ~is_nan);`
try to use use v128 as well as it has builtin operator for these.
nans can be constructed as v128 using v128::from32p(ppu_man_f32).

Comment (VelocityRa) [reply rpcs3/Emu/Cell/PPUInterpreter.cpp:372]: nice, thx

File: rpcs3/Emu/Cell/PPUInterpreter.cpp [hunk 4]
```diff
@@ -358,6 +359,33 @@ class ppu_scale_table_t
...
 
+constexpr u32 ppu_nan_u32 = 0x7FC00000u;
+const f32 ppu_nan_f32 = std::bit_cast<f32>(ppu_nan_u32);
+
+template<typename... Args>
+v128 vec_handle_nan(v128 result, Args... args)
+{
+	const auto is_nan = v128::fromF(_mm_cmpunord_ps(result.vf, result.vf));
+	const auto nans = v128::from32p(ppu_nan_u32);
+	result = (nans & is_nan) | (result & ~is_nan);
```
Comment (elad335) [inline rpcs3/Emu/Cell/PPUInterpreter.cpp:370]:
```suggestion
	result = (nans & is_nan) | v128::andnot(result, is_nan);
```
tiny optimization.

File: rpcs3/Emu/Cell/PPUInterpreter.cpp [hunk 5]
```diff
@@ -1906,7 +1948,10 @@ bool ppu_interpreter::VRLW(ppu_thread& ppu, ppu_opcode_t op)
 
 bool ppu_interpreter::VRSQRTEFP(ppu_thread& ppu, ppu_opcode_t op)
 {
-	ppu.vr[op.vd].vf = _mm_div_ps(_mm_set_ps(1.0f, 1.0f, 1.0f, 1.0f), _mm_sqrt_ps(ppu.vr[op.vb].vf));
+	const auto a = _mm_set_ps(1.0f, 1.0f, 1.0f, 1.0f);
+	const auto b = _mm_sqrt_ps(ppu.vr[op.vb].vf);
```
Comment (elad335) [inline rpcs3/Emu/Cell/PPUInterpreter.cpp:1952]: the argument passed to vec_handle_nan needs to be unmodified vb.

File: rpcs3/Emu/Cell/PPUTranslator.h [hunk 6]
```diff
@@ -99,6 +101,14 @@ class PPUTranslator final : public cpu_translator
 		return SetVr(vr, expr.eval(m_ir));
 	}
 
+	llvm::Value* VecHandleNan(llvm::Value* val);
+
+	template <typename T>
+	auto vec_handle_nan(T&& expr)
+	{
+		return VecHandleNan(expr.eval(m_ir));
```
Comment (elad335) [inline rpcs3/Emu/Cell/PPUTranslator.h:109]: maybe return a value_t<> instead of llvm::Value* for consistency.

**Comment by Fneufneu** [comment]
Tested with the windows build from azure and i have some glitchs in the game "Journey"

OpenGL: https://youtu.be/mrlOdhW978U
Vulkan: https://youtu.be/xadBhYEDFG8

Vulkan video without this PR: https://youtu.be/-V6MuMdOrEE

**Comment by elad335** [comment]
note for testers: clear ppu game cache and firmware cache.

**Comment by Yahfz** [comment]
> Tested with the windows build from azure and i have some glitchs in the game "Journey"
> Intel(R) Core(TM) i5-8600K CPU @ 3.60GHz | 6 Threads | 15.94 GiB RAM | TSC: 3.600GHz | AVX+ | FMA3 | TSX
> VEGA 64
> 
> Vulkan: https://youtu.be/xadBhYEDFG8
> Vulkan video without this PR: https://youtu.be/-V6MuMdOrEE

Thats unrelated and happens when the cpu isn't fast enough.

**Comment by Fneufneu** [comment]
ok, i forgot to clear firmware cache but i still get around 18 fps with this PR and 19 fps without.
and lot's of glitchs in menu and in game vs 0 glitch.
overlay max fps: with PR: 19.3 and without 20.6

File: rpcs3/Emu/Cell/PPUTranslator.h (hunk 6 above)
Comment (VelocityRa) [reply rpcs3/Emu/Cell/PPUTranslator.h:109]: i considered it but `set_vr` above this is the same style and does _not_ return `value_t`, so the _current_ way is consistent with that

Comment (elad335) [reply rpcs3/Emu/Cell/PPUTranslator.h:109]: I mean its consistent with other functions in CPUTranslator.h
It's also odd to mix "old style" llvm::Value* functions with the "new style" value_t<>.

Comment (elad335) [reply rpcs3/Emu/Cell/PPUTranslator.h:109]: set_vr also doesn't return anything so it's not a fair comparison.

Comment (VelocityRa) [reply rpcs3/Emu/Cell/PPUTranslator.h:109]: ah yeah, does call `eval` but doesn't return, i'll change it

Comment (VelocityRa) [reply rpcs3/Emu/Cell/PPUTranslator.h:109]: wait i'm not crazy it does return lol
but ret type is `void` 🤔 
```
	template <typename T>
	void set_vr(u32 vr, T&& expr)
	{
		return SetVr(vr, expr.eval(m_ir));
	}
```

Comment (VelocityRa) [reply rpcs3/Emu/Cell/PPUTranslator.h:109]: i'll change it

Comment (VelocityRa) [reply rpcs3/Emu/Cell/PPUTranslator.h]: pushed

**Comment by Nekotekina** [comment]
I think it should be under an option.

**Comment by VelocityRa** [comment]
Pushed (not exposed to GUI).

**Comment by Nekotekina** [comment]
Hmm, not pushed?

**Comment by elad335** [comment]
The option needs to affect ppu cache as well. (see accurate ppu dfma setting handling)

**Comment by VelocityRa** [comment]
pushed

**Comment by rafael-57** [comment]
This doesn't seem to fix issue #5289, is the option enabled by default?

**Comment by AniLeo** [comment]
No

**Comment by rafael-57** [comment]
How do we set it? Or do we need to wait for a PR to expose it to the GUI?

> 
> 
> No

**Comment by RainbowCookie32** [comment]
You have to set `PPU LLVM Accurate Vector NaN values` to `true` in config.yml until it's exposed to the GUI. You might also have to clear PPU Cache as well, not sure about that

**Comment by elad335** [comment]
or use ppu interpreter.

**Comment by rafael-57** [comment]
I haven't been able to set the option in the config .yml for some reason, but this does fix  #5289 when using ppu interpreter

**Pull request merged**